### PR TITLE
Adding a way to subscribe to circulars for email notificaions

### PR DESCRIPTION
**Pull request by dakota002**

### CODE REVIEW

**Comment by lpsinger** [comment]
Would you please post some screen shots and recordings to walk me through the UX?

**Comment by jracusin** [comment]
Default should be to be subscribed, not unsubscribed.

**Comment by jracusin** [comment]
The current page is kind of confusing between notices and circulars.  Can users add multiple circulars email subscription with the Add button?

Add a separate subsection for Notices, like you have for circulars.

**Comment by dakota002** [comment]
@jracusin  The subsection exists, but doesnt show anything unless the user creates an alert
image

As for defaulting the user to be subscribed, I think that is outside of the scope of this PR. For new users that would involve modifying the post confirm signup function. And for existing users it will probably be easiest to do that with a script to update the table programmatically

**Comment by jracusin** [comment]
@dakota002, I'm suggesting something like the following:

Create and manage email subscriptions to GCN Circulars and Notices here.

**GCN Circulars**
Text you already have.
<strong>Circulars</strong> are sent from GCN Circulars{' '}
         {`<no-reply@${hostname}>`} and will be delivered to the email associated
         with this account. 

**GCN Notices**
         You can create as many <strong>Notice</strong>{' '}
         subscription alerts as you like. To create a new alert, click the "Add"
         button. Once an alert has been created, you can click the "Test Message"
         button to send a test email to the listed recipient, to verify that the
         emails will make it into your inbox.

Add button

**Comment by dakota002** [comment]
Sure! Give me a few, and I will post some pictures of an updated version

**Comment by jracusin** [comment]
Yeah, that looks good.  Though either remove "Subscriptions" from "GCN Notice Subscriptions" or add it to "GCN Circulars", just so they're consistent language.

File: app/routes/user/email/index.tsx [hunk 1]
```diff
@@ -70,7 +86,55 @@ export default function () {
...
+            You are not currently subscribed to receive GCN Circulars via Email.
+          </p>
+          <Form method="post">
+            <input type="hidden" name="intent" value={'subscribe'} />
+            <Button type={'submit'} unstyled>
+              Subscribe
+            </Button>
+          </Form>
+        </>
+      )}
```
Comment (lpsinger) [inline app/routes/user/email/index.tsx:119]: We can do better than this for the UX. I propose:

- [ ] A [segmented button group](https://trussworks.github.io/react-uswds/?path=/story/components-button-groups--segmented) with buttons `On` and `Off`, mimicking a toggle button.
- [ ] A spinner to the right of the button group that displays while the form is submitting (we have several examples of these in our code now)
- [ ] A green checkmark and the text `Saved` that displays to the right of the button group when the form submission has completed (we also have several examples of these)

@dakota002, is your Remix-fu up to the task?

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: app/routes/user/email/index.tsx (hunk 1 above)
Comment (dakota002) [reply app/routes/user/email/index.tsx:119]: Challenge accepted!

Comment (lpsinger) [reply app/routes/user/email/index.tsx:119]: 🥋 💪

Comment (dakota002) [reply app/routes/user/email/index.tsx:119]: image

image

image

File: app/routes/user/email_circulars.server.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
+import { tables } from '@architect/functions'
+
+export async function createGcnCircularEmailNotification(
```
Comment (lpsinger) [inline app/routes/user/email_circulars.server.ts:3]: Remove `Gcn` from all method names.

```suggestion
export async function createCircularEmailNotification(
```

File: app/routes/user/email/index.tsx [hunk 3]
```diff
@@ -7,53 +7,79 @@
...
       }
     case 'sendTest':
       const recipient = getFormDataString(data, 'recipient')
       if (recipient) {
-        const machine = await EmailNotificationServer.create(request)
-        await machine.sendTestEmail(recipient)
+        await sendTestEmail(recipient)
       }
+    case 'subscribe':
+      await createGcnCircularEmailNotification(user?.sub, user?.email)
```
Comment (lpsinger) [inline app/routes/user/email/index.tsx:52]: Did TypeScript not infer that user must be defined from the `if (!user) throw ...` above?

```suggestion
      await createGcnCircularEmailNotification(user.sub, user.email)
```

File: app/routes/user/email/index.tsx [hunk 4]
```diff
@@ -7,53 +7,79 @@
...
+      break
+    case 'unsubscribe':
+      await deleteGcnCircularEmailNotification(user?.sub, user?.email)
```
Comment (lpsinger) [inline app/routes/user/email/index.tsx:55]:
```suggestion
      await deleteGcnCircularEmailNotification(user.sub, user.email)
```

File: app/routes/user/email/index.tsx [hunk 5]
```diff
@@ -7,53 +7,79 @@
...
 export async function loader({ request }: DataFunctionArgs) {
-  const machine = await EmailNotificationServer.create(request)
-  const data = await machine.getEmailNotifications()
-  return data
+  const user = await getUser(request)
+  if (!user) throw new Response(null, { status: 403 })
+  const data = await getEmailNotifications(user.sub)
+  const hostname = getHostname()
+  const userIsSubscribedToCircularEmails =
+    await getUsersCircularSubmissionStatus(user?.sub)
```
Comment (lpsinger) [inline app/routes/user/email/index.tsx:67]:
```suggestion
    await getUsersCircularSubmissionStatus(user.sub)
```

File: app/routes/user/email/index.tsx [hunk 6]
```diff
@@ -70,7 +96,55 @@ export default function () {
         </a>
         .
       </p>
-      {data && (
+      <h2>GCN Circulars</h2>
+      <p>
+        Subscriptions to <strong>Circulars</strong> are sent from GCN Circulars{' '}
+        {`<no-reply@${hostname}>`} and will be delivered to the email associated
+        with this account.
```
Comment (lpsinger) [inline app/routes/user/email/index.tsx:103]: Add the user's email address here.

```suggestion
        with your account ({email}).
```

File: app/routes/user/email/index.tsx [hunk 7]
```diff
@@ -70,7 +96,55 @@ export default function () {
         </a>
         .
       </p>
-      {data && (
+      <h2>GCN Circulars</h2>
+      <p>
+        Subscriptions to <strong>Circulars</strong> are sent from GCN Circulars{' '}
+        {`<no-reply@${hostname}>`} and will be delivered to the email associated
```
Comment (lpsinger) [inline app/routes/user/email/index.tsx:102]: Tense agreement:

```suggestion
        {`<no-reply@${hostname}>`} and are delivered to the email associated
```

File: app/routes/user/email/index.tsx [hunk 8]
```diff
@@ -70,7 +96,55 @@ export default function () {
         </a>
         .
       </p>
-      {data && (
+      <h2>GCN Circulars</h2>
```
Comment (lpsinger) [inline app/routes/user/email/index.tsx:99]:
```suggestion
      <h2>Circulars</h2>
```

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: app/components/HeadingWithAddButton.tsx [hunk 9]
```diff
@@ -9,17 +9,24 @@ import { Link } from '@remix-run/react'
...
   return (
     <Grid row>
       <Grid tablet={{ col: 'fill' }}>
-        <h1>{children}</h1>
+        <CustomTag>{children}</CustomTag>
       </Grid>
       <Grid tablet={{ col: 'auto' }}>
-        <Link className="usa-button tablet:margin-right-2" to="edit">
+        <Link
+          className="usa-button tablet:margin-right-2 tablet:margin-top-105"
```
[the remark this answers is not in the document]
Comment (dakota002) [reply app/components/HeadingWithAddButton.tsx:27]: It adds some whitespace at the top of the button to get the alignment looking better. Without it, the button alignment looks weird:
image

**Comment by lpsinger** [comment]
This all needs to be hidden behind the feature flag!

**Review by lpsinger — CHANGES_REQUESTED** [review]
The circulars settings and the actions need to be hidden behind the feature flag.

File: app/routes/user/email/index.tsx [hunk 10]
```diff
@@ -74,12 +75,14 @@ export default function () {
...
   const [dirty, setDirty] = useState(false)
+  const enableCirculars = useFeature('circulars')
 
   return (
     <>
       <h1>Email Notifications</h1>
       <p className="usa-paragraph">
-        Create and manage email subscriptions to GCN Circulars and Notices here.
+        Create and manage email subscriptions to{' '}
+        {enableCirculars && 'GCN Circulars and'} Notices here.
```
Comment (lpsinger) [inline app/routes/user/email/index.tsx:85]:
```suggestion
        Create and manage email subscriptions to GCN{' '}
        {enableCirculars && 'Circulars and'} Notices here.
```

**Review by lpsinger — CHANGES_REQUESTED** [review]
Also lock out the new action function logic with the feature flag.

File: app/routes/user/email/index.tsx [hunk 11]
```diff
@@ -7,53 +7,139 @@
...
       }
     case 'sendTest':
       const recipient = getFormDataString(data, 'recipient')
       if (recipient) {
-        const machine = await EmailNotificationServer.create(request)
-        await machine.sendTestEmail(recipient)
+        await sendTestEmail(recipient)
       }
+    case 'subscribe':
+      if (!feature('circulars')) break
```
Comment (lpsinger) [inline app/routes/user/email/index.tsx:52]:
```suggestion
      if (!feature('circulars')) throw new Response(null, { status: 400 })
```

**Review by lpsinger — CHANGES_REQUESTED** [review]
These buttons are arranged confusingly. It's not visually clear that they don't belong to the same form. How we can separate them more clearly?

Screenshot 2023-04-04 at 12 17 23

File: app/components/HeadingWithAddButton.tsx [hunk 12]
```diff
@@ -9,17 +9,27 @@ import { Link } from '@remix-run/react'
...
-        <h1>{children}</h1>
+        <CustomTag>{children}</CustomTag>
       </Grid>
       <Grid tablet={{ col: 'auto' }}>
-        <Link className="usa-button tablet:margin-right-2" to="edit">
+        <Link
+          className={
+            'usa-button tablet:margin-right-2' +
+            (headingLevel > 1 ? ' tablet:margin-top-105' : '')
+          }
```
Comment (lpsinger) [inline app/components/HeadingWithAddButton.tsx:30]:
```suggestion
          className={
            classNames('usa-button', 'tablet:margin-right-2', {'tablet:margin-top-105': headingLevel > 1})
      }
```

**Review by lpsinger — CHANGES_REQUESTED** [review]

**Review by lpsinger — APPROVED** [review]

**Pull request merged**